docs/rfcs/20201208_fragmentation.md

```diff
@@ -0,0 +1,303 @@
+
+## Mitigation Design
+
+Here we examine several possible solutions to the fragmentation problem.
+
+We broadly group them into avoidance, allocator modifications and application (redpanda) modifications.
+
+### Avoid
+
+Broadly speaking, the solution here is: *don't try to avoid fragmentation, but rather try to avoid failure even when fragmentation occurs*. The main approach is to simply to _avoid large allocations_.
+
+This is the same approach taken by the Linux kernel (which uses a buddy allocator) for `kmalloc` (the guideline is expect allocators greater than order 4 to fail and for a long time the maximum allocation size was 128 KiB) as well as ScyllaDB (see for example issues [#7364](https://github.com/scylladb/scylla/issues/7364), [#6376](https://github.com/scylladb/scylla/issues/6376), [#108](https://github.com/scylladb/scylla/issues/108), [#1546](https://github.com/scylladb/scylla/issues/1546) and many more).
+
+The main tools to _detect_ such cases are using the large allocation warning emitted by Seastar (not currently configured in Redpanda), and the diagnostics when an allocation fails. 
+
+Once a large allocation is detected it, avoiding it depends on the context (browsing the related ScyllaDB issues shows the large variety of approaches taken). Sometimes it is as simple as using a fragmented data structure (linked list, iobuf, deque) rather than a contiguous one. In other cases it was not expected that such a structure became so large in the first place, so some other adjustment is needed.
+
+#### Drawbacks
+
+A significant downside of the this approach is that is mostly reactive: problems are identified only when they cause a crash or (hopefully) when the warning is detected and acted on. Although the issue may be resolved, significant downtime may already have been incurred.
+
+Another issue is that many allocations occur in third party libraries rather than the core code: although we have the source to all third parties and can patch them, this creates a divergence with upstream, creates a maintenance burden, and so on. Furthermore, it may be harder to avoid large allocations in third party libraries since we won't the same tools available to use as 
+
+#### Recommendations
+
+At least, we should enable the large allocation warning, perhaps at 256 KiB. We can have it enabled in testing with a post-processing phase which determines if the discovered large allocations are from new locations which need to vetted to check how large they can get.
+
+We should expose metrics related to fragmentation such as availability % for certain sizes in our metrics and possibly in telemetry, in order to understand what "typical" fragmentation looks like.
```

Comment:
Is the availability percentage the patch you created but isn't merged upstream? If so, perhaps we should poke the mailing list again and also commit into our fork?

Reply:
@dotnwat - that's right, it's only on my local box. I have a fair amount of random WIP there that isn't upstream (but some of the more serious stuff is already upstream). 

I certainly could sort it out and push it upstream, just a matter of prioritization (the availability % part is very simple).